Optimize pointer to matrix assignment

facug91 opened the pull request:
As discussed [here](https://github.com/davisking/dlib/discussions/2814), this assignment could be optimized, and that't what this PR does.  
Here, I also check for data size to be different. If it's not, it doesn't call set_size, which make the assignment even faster that the benchmarks I've said in the discussion.  
I've implemented it with op_pointer_to_mat as Davis suggested, which is great because now everyone who already uses this method will have an efficiency improvement just by upgrading the library.

facug91 commented:
Wow, that's a lot of bugs. Well, I'll start going through them now, sorry.

pfeatherstone commented:
I've reproduced your tests and `mat()` is just as performant as your suggested change. So i don't know why you're getting different answers.

pfeatherstone commented:
I used the following:
```
#include <opencv2/opencv.hpp>
#include <dlib/matrix.h>
#include <dlib/opencv.h>
#include <dlib/image_transforms/assign_image.h>

using namespace dlib;
using namespace std::chrono;

void assign_ptr(matrix<unsigned char>& dst, const cv_image<unsigned char>& src) {
	dst.set_size(src.nr(), src.nc());
	auto* d    = &dst(0,0);
	auto* aux  = &src(0,0);
	for (long i=0; i<src.size(); ++i)
		*d++ = *aux++;
}

int main(int argc, char** argv) 
{
	cv::VideoCapture cap("/home/pf/Downloads/repos/mat/build/_deps/dlib-src/dlib/test/ffmpeg_data/MOT17-13-SDP-raw.h265");
	cv::Mat frame, gray;
	cap >> frame;
    cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);

	matrix<unsigned char> aux(gray.rows, gray.cols);
    
    const int ntests = 500000;

    {
        const auto start = high_resolution_clock::now();
        for (int i = 0 ; i < ntests ; ++i)
            assign_image(aux, dlib::cv_image<unsigned char>(gray));
        const long ans = sum(aux);
        std::cout << "assign_image : " << duration_cast<microseconds>(high_resolution_clock::now() - start).count() / ntests << " us" << std::endl;
        std::cout << "Sum of mats  : " << ans << std::endl;
    }
    
    {
        const auto start = high_resolution_clock::now();
        for (int i = 0 ; i < ntests ; ++i)
            aux = mat(gray.data, gray.rows, gray.cols);
        const long ans = sum(aux);
        std::cout << "mat          : " << duration_cast<microseconds>(high_resolution_clock::now() - start).count() / ntests << " us" << std::endl;
        std::cout << "Sum of mats  : " << ans << std::endl;
    }

    {
        const auto start = high_resolution_clock::now();
        for (int i = 0 ; i < ntests ; ++i)
            assign_ptr(aux, gray);
        const long ans = sum(aux);
        std::cout << "assign_ptr   : " << duration_cast<microseconds>(high_resolution_clock::now() - start).count() / ntests << " us" << std::endl;
        std::cout << "Sum of mats  : " << ans << std::endl;
    }

	return 0;
}
```
On my machine, I get:
```
assign_image : 9 us
Sum of mats  : 128
mat          : 9 us
Sum of mats  : 128
assign_ptr   : 9 us
Sum of mats  : 128
```

pfeatherstone commented:
Are you sure you built your code in Release mode?

pfeatherstone commented:
In Debug mode, I get:
```
assign_image : 5180 us
Sum of mats  : 128
mat          : 6451 us
Sum of mats  : 128
assign_ptr   : 916 us
Sum of mats  : 128
```
That seems similar to what you got. I'm willing to bet you built in Debug mode. C++ has really poor performance in Debug.

davisking commented:
Yeah definitely double check that you built with optimizations.  Also, upon second thought, the thing to overload that makes avoiding inclusion order problems is `matrix_assign()`.  See how that function is used in the `matrix` class.  It's the function that catches and does all the smart overloads like this.  An overload of `matrix_assign(matrix_exp<matrix_op<op_pointer_to_mat<T>>>)` could I think be placed right next to `mat()` and it would compile and do what you want here.

Although yeah, definitely check that optimizations were on first :)

pfeatherstone commented:
Doesn't `mat()` do exactly what we want anyway? Unless I'm missing something here. I can't quite see the missing optimization.

davisking commented:
> Doesn't `mat()` do exactly what we want anyway? Unless I'm missing something here. I can't quite see the missing optimization.

Maybe, depends on how clever the optimizer is in the compiler.  But for instance there is a multiplication by stride which maybe doesn't get optimized out. Hard to say.  Just have to test it and see what happens.

pfeatherstone commented:
> > Doesn't `mat()` do exactly what we want anyway? Unless I'm missing something here. I can't quite see the missing optimization.
> 
> Maybe, depends on how clever the optimizer is in the compiler. But for instance there is a multiplication by stride which maybe doesn't get optimized out. Hard to say. Just have to test it and see what happens.

Ah I see. Yeah maybe an optimization using memcpy() where appropriate would be good. Well, in my tests I get the exact same result. Maybe it's gcc 12.1 being really clever.

facug91 commented:
I wasn't using Debug mode, I know C++ code without optimization is unusable. There's still something embarrassing though, which is that I was using RelWithDebInfo for the profiler, which I thought would have the same performance as Release (because it's Release With Debug Information), but seems to be much less optimized than Release.

I've changed the method `assign_ptr` in your script with:
```cpp
void assign_ptr(matrix<unsigned char>& dst, const cv_image<unsigned char>& src) {
	if (dst.nr() != src.nr() || dst.nc() == src.nc())
		dst.set_size(src.nr(), src.nc());
	const long size = dst.nr() * dst.nc();
	auto* d    = &dst(0,0);
	auto* aux  = &src(0,0);
	for (long i=0; i<size; ++i)
		*d++ = *aux++;
}
```
And try it with that video, and with another one that is full hd, and this new method is still faster than the previous ones, albeit only ~9%. I don't know if it worth it.  

Frame resolution: [960 x 540]
assign_image : 22724 ns
Sum of mats  : 128
mat          : 22532 ns
Sum of mats  : 128
assign_ptr   : 20403 ns
Sum of mats  : 128
~9.45% faster than mat

Frame resolution: [1920 x 1080]
assign_image : 94763 ns
Sum of mats  : 102
mat          : 89276 ns
Sum of mats  : 102
assign_ptr   : 80933 ns
Sum of mats  : 102
~9.35% faster than mat

Sorry for the confusion.

davisking commented:
> I wasn't using Debug mode, I know C++ code without optimization is unusable. There's still something embarrassing though, which is that I was using RelWithDebInfo for the profiler, which I thought would have the same performance as Release (because it's Release With Debug Information), but seems to be much less optimized than Release.
> 
> I've changed the method `assign_ptr` in your script with:
> 
> ```c++
> void assign_ptr(matrix<unsigned char>& dst, const cv_image<unsigned char>& src) {
> 	if (dst.nr() != src.nr() || dst.nc() == src.nc())
> 		dst.set_size(src.nr(), src.nc());
> 	const long size = dst.nr() * dst.nc();
> 	auto* d    = &dst(0,0);
> 	auto* aux  = &src(0,0);
> 	for (long i=0; i<size; ++i)
> 		*d++ = *aux++;
> }
> ```
> 
> And try it with that video, and with another one that is full hd, and this new method is still faster than the previous ones, albeit only ~9%. I don't know if it worth it.
> 
> Frame resolution: [960 x 540] assign_image : 22724 ns Sum of mats : 128 mat : 22532 ns Sum of mats : 128 assign_ptr : 20403 ns Sum of mats : 128 ~9.45% faster than mat
> 
> Frame resolution: [1920 x 1080] assign_image : 94763 ns Sum of mats : 102 mat : 89276 ns Sum of mats : 102 assign_ptr : 80933 ns Sum of mats : 102 ~9.35% faster than mat
> 
> Sorry for the confusion.

No worries.  Try the `matrix_assign()` overload way I mentioned.  And if you have an overload specifically for trivially copyable types that uses memcpy() it might be even faster.

pfeatherstone commented:
![tesla](https://github.com/davisking/dlib/assets/45853521/cb6eddd9-3073-491f-a917-199d4bae3cb8)

pfeatherstone commented:
```
assign_image : 333 us
Sum of mats  : 65
mat          : 337 us
Sum of mats  : 65
assign_ptr   : 441 us
Sum of mats  : 65
```

facug91 commented:
> > I wasn't using Debug mode, I know C++ code without optimization is unusable. There's still something embarrassing though, which is that I was using RelWithDebInfo for the profiler, which I thought would have the same performance as Release (because it's Release With Debug Information), but seems to be much less optimized than Release.
> > I've changed the method `assign_ptr` in your script with:
> > ```c++
> > void assign_ptr(matrix<unsigned char>& dst, const cv_image<unsigned char>& src) {
> > 	if (dst.nr() != src.nr() || dst.nc() == src.nc())
> > 		dst.set_size(src.nr(), src.nc());
> > 	const long size = dst.nr() * dst.nc();
> > 	auto* d    = &dst(0,0);
> > 	auto* aux  = &src(0,0);
> > 	for (long i=0; i<size; ++i)
> > 		*d++ = *aux++;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And try it with that video, and with another one that is full hd, and this new method is still faster than the previous ones, albeit only ~9%. I don't know if it worth it.
> > Frame resolution: [960 x 540] assign_image : 22724 ns Sum of mats : 128 mat : 22532 ns Sum of mats : 128 assign_ptr : 20403 ns Sum of mats : 128 ~9.45% faster than mat
> > Frame resolution: [1920 x 1080] assign_image : 94763 ns Sum of mats : 102 mat : 89276 ns Sum of mats : 102 assign_ptr : 80933 ns Sum of mats : 102 ~9.35% faster than mat
> > Sorry for the confusion.
> 
> No worries. Try the `matrix_assign()` overload way I mentioned. And if you have an overload specifically for trivially copyable types that uses memcpy() it might be even faster.

I dind't try the `matrix_assign()` yet, but I did try replacing the loop with `memcpy()`, and it is much faster. Although, I don't know why my results are so different than yours @pfeatherstone .  Apparently your CPU is better, because it runs faster, and also I'm using gcc 9.4, instead of gcc 12.1. But I don't think that should make such differences.

I used these images downloaded from Google:
1280x720.png:
![1280x720](https://github.com/davisking/dlib/assets/7832961/72bfff49-c916-441e-9bec-772838d7404b)
1920x1080.jpg:
![1920x1080](https://github.com/davisking/dlib/assets/7832961/57a274a1-daf4-4c0d-8feb-5f81eb4ece6d)
4685x2143.jpg:
(The same as the comment above).
6720x4480.jpg:
![6720x4480](https://github.com/davisking/dlib/assets/7832961/6b53320e-79ce-4e23-8add-0017a7b51d6c)

I modified the script to use `memcpy()` and to execute all resolutions at the same time:
```cpp
#include <opencv2/opencv.hpp>
#include <dlib/matrix.h>
#include <dlib/opencv.h>
#include <dlib/image_transforms/assign_image.h>

using namespace dlib;
using namespace std::chrono;

void assign_ptr(matrix<unsigned char>& dst, const cv_image<unsigned char>& src) {
	if (dst.nr() != src.nr() || dst.nc() == src.nc())
		dst.set_size(src.nr(), src.nc());
	const long size = dst.nr() * dst.nc();
	std::memcpy(&dst(0,0), &src(0,0), size);
}

int main(int argc, char** argv)
{
	std::vector<std::string> imagePaths = {
			"/home/investigacion/computer-vision/1280x720.png",
			"/home/investigacion/computer-vision/1920x1080.jpg",
			"/home/investigacion/computer-vision/4685x2143.jpg",
			"/home/investigacion/computer-vision/6720x4480.jpg"
	};
	for (const auto& imagePath : imagePaths) {
		cv::Mat frame = cv::imread(imagePath), gray;
		cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
		std::cout << "Frame resolution: " << gray.size() << std::endl;

		matrix<unsigned char> aux(gray.rows, gray.cols);

		const int ntests = 50000;

		{
			const auto start = high_resolution_clock::now();
			for (int i = 0; i < ntests; ++i)
				assign_image(aux, dlib::cv_image<unsigned char>(gray));
			const long ans = sum(aux);
			std::cout << "assign_image : " << duration_cast<nanoseconds>(high_resolution_clock::now() - start).count() / ntests << " ns"
			          << std::endl;
			std::cout << "Sum of mats  : " << ans << std::endl;
		}

		{
			const auto start = high_resolution_clock::now();
			for (int i = 0; i < ntests; ++i)
				aux = mat(gray.data, gray.rows, gray.cols);
			const long ans = sum(aux);
			std::cout << "mat          : " << duration_cast<nanoseconds>(high_resolution_clock::now() - start).count() / ntests << " ns"
			          << std::endl;
			std::cout << "Sum of mats  : " << ans << std::endl;
		}

		{
			const auto start = high_resolution_clock::now();
			for (int i = 0; i < ntests; ++i)
				assign_ptr(aux, gray);
			//aux.assign_ptr(gray.data, gray.rows, gray.cols);
			const long ans = sum(aux);
			std::cout << "assign_ptr   : " << duration_cast<nanoseconds>(high_resolution_clock::now() - start).count() / ntests << " ns"
			          << std::endl;
			std::cout << "Sum of mats  : " << ans << std::endl;
		}
	}

	return 0;
}
```

With the previous implementation (without using `memcpy()`), these were the results:

Frame resolution: [1280 x 720]
assign_image : 43517 ns
Sum of mats  : 21
mat          : 43191 ns
Sum of mats  : 21
assign_ptr   : 36789 ns
Sum of mats  : 21
~14.82% faster than mat

Frame resolution: [1920 x 1080]
assign_image : 90866 ns
Sum of mats  : 56
mat          : 90991 ns
Sum of mats  : 56
assign_ptr   : 85773 ns
Sum of mats  : 56
~5.73% faster than mat

Frame resolution: [4685 x 2143]
assign_image : 871835 ns
Sum of mats  : 17
mat          : 893350 ns
Sum of mats  : 17
assign_ptr   : 831742 ns
Sum of mats  : 17
~6.9% faster than mat

Frame resolution: [6720 x 4480]
assign_image : 3015114 ns
Sum of mats  : 199
mat          : 3009109 ns
Sum of mats  : 199
assign_ptr   : 2901399 ns
Sum of mats  : 199
~3.6% faster than mat

With `memcpy()` implementation:

Frame resolution: [1280 x 720]
assign_image : 42956 ns
Sum of mats  : 21
mat          : 44065 ns
Sum of mats  : 21
assign_ptr   : 30934 ns
Sum of mats  : 21
~29.8% faster than mat

Frame resolution: [1920 x 1080]
assign_image : 90384 ns
Sum of mats  : 56
mat          : 90048 ns
Sum of mats  : 56
assign_ptr   : 71866 ns
Sum of mats  : 56
~20.2% faster than mat

Frame resolution: [4685 x 2143]
assign_image : 882964 ns
Sum of mats  : 17
mat          : 873336 ns
Sum of mats  : 17
assign_ptr   : 564575 ns
Sum of mats  : 17
~35.4% faster than mat

Frame resolution: [6720 x 4480]
assign_image : 3102560 ns
Sum of mats  : 199
mat          : 3026419 ns
Sum of mats  : 199
assign_ptr   : 2045988 ns
Sum of mats  : 199
~32.4% faster than mat

pfeatherstone commented:
I've added `matrix_assign()` and removes `sum(aux)` since it overflows and gives gibberish anyway. 
```
#include <opencv2/opencv.hpp>
#include <dlib/matrix.h>
#include <dlib/opencv.h>
#include <dlib/image_transforms/assign_image.h>

using namespace dlib;
using namespace std::chrono;

void assign_ptr(matrix<unsigned char>& dst, const cv_image<unsigned char>& src) 
{
	dst.set_size(src.nr(), src.nc());
	const long size = dst.nr() * dst.nc();
    std::memcpy(&dst(0,0), &src(0,0), size);
}

int main(int argc, char** argv) 
{
	cv::Mat gray        = cv::imread("/home/pf/Downloads/tesla2.jpg", cv::IMREAD_GRAYSCALE);
    const auto dgray    = dlib::cv_image<unsigned char>(gray);
    std::cout << "Dims " << gray.rows << "x" << gray.cols << std::endl;

	matrix<unsigned char> aux(gray.rows, gray.cols);
    
    const int ntests = 1000;

    {
        const auto start = high_resolution_clock::now();
        for (int i = 0 ; i < ntests ; ++i)
            matrix_assign(aux, mat(dgray));
        std::cout << "matrix_assign: " << duration_cast<nanoseconds>(high_resolution_clock::now() - start).count() / ntests << " ns" << std::endl;
    }

    {
        const auto start = high_resolution_clock::now();
        for (int i = 0 ; i < ntests ; ++i)
            assign_image(aux, dgray);
        std::cout << "assign_image : " << duration_cast<nanoseconds>(high_resolution_clock::now() - start).count() / ntests << " ns" << std::endl;
    }
    
    {
        const auto start = high_resolution_clock::now();
        for (int i = 0 ; i < ntests ; ++i)
            aux = mat(gray.data, gray.rows, gray.cols);
        std::cout << "mat          : " << duration_cast<nanoseconds>(high_resolution_clock::now() - start).count() / ntests << " ns" << std::endl;
    }

    {
        const auto start = high_resolution_clock::now();
        for (int i = 0 ; i < ntests ; ++i)
            assign_ptr(aux, gray);
        std::cout << "assign_ptr   : " << duration_cast<nanoseconds>(high_resolution_clock::now() - start).count() / ntests << " ns" << std::endl;
    }

	return 0;
}
```
I used the biggest image and get:
```
Dims 4480x6720
matrix_assign: 2209512 ns
assign_image : 2215667 ns
mat          : 2275655 ns
assign_ptr   : 1572658 ns
```
So yeah, with really big images, `memcpy()` starts to come through on top.

facug91 commented:
> I've added `matrix_assign()` and removed `sum(aux)` since it overflows and gives gibberish anyway.
> 
> ```
> #include <opencv2/opencv.hpp>
> #include <dlib/matrix.h>
> #include <dlib/opencv.h>
> #include <dlib/image_transforms/assign_image.h>
> 
> using namespace dlib;
> using namespace std::chrono;
> 
> void assign_ptr(matrix<unsigned char>& dst, const cv_image<unsigned char>& src) 
> {
> 	dst.set_size(src.nr(), src.nc());
> 	const long size = dst.nr() * dst.nc();
>     std::memcpy(&dst(0,0), &src(0,0), size);
> }
> 
> int main(int argc, char** argv) 
> {
> 	cv::Mat gray        = cv::imread("/home/pf/Downloads/tesla2.jpg", cv::IMREAD_GRAYSCALE);
>     const auto dgray    = dlib::cv_image<unsigned char>(gray);
>     std::cout << "Dims " << gray.rows << "x" << gray.cols << std::endl;
> 
> 	matrix<unsigned char> aux(gray.rows, gray.cols);
>     
>     const int ntests = 1000;
> 
>     {
>         const auto start = high_resolution_clock::now();
>         for (int i = 0 ; i < ntests ; ++i)
>             matrix_assign(aux, mat(dgray));
>         std::cout << "matrix_assign: " << duration_cast<nanoseconds>(high_resolution_clock::now() - start).count() / ntests << " ns" << std::endl;
>     }
> 
>     {
>         const auto start = high_resolution_clock::now();
>         for (int i = 0 ; i < ntests ; ++i)
>             assign_image(aux, dgray);
>         std::cout << "assign_image : " << duration_cast<nanoseconds>(high_resolution_clock::now() - start).count() / ntests << " ns" << std::endl;
>     }
>     
>     {
>         const auto start = high_resolution_clock::now();
>         for (int i = 0 ; i < ntests ; ++i)
>             aux = mat(gray.data, gray.rows, gray.cols);
>         std::cout << "mat          : " << duration_cast<nanoseconds>(high_resolution_clock::now() - start).count() / ntests << " ns" << std::endl;
>     }
> 
>     {
>         const auto start = high_resolution_clock::now();
>         for (int i = 0 ; i < ntests ; ++i)
>             assign_ptr(aux, gray);
>         std::cout << "assign_ptr   : " << duration_cast<nanoseconds>(high_resolution_clock::now() - start).count() / ntests << " ns" << std::endl;
>     }
> 
> 	return 0;
> }
> ```
> 
> I used the biggest image and get:
> 
> ```
> Dims 4480x6720
> matrix_assign: 2209512 ns
> assign_image : 2215667 ns
> mat          : 2275655 ns
> assign_ptr   : 1572658 ns
> ```
> 
> So yeah, with really big images, `memcpy()` starts to come through on top. So it looks like the compiler isn't able to optimize through the expression template. That's annoying.

With large images, and/or maybe with less powerful cpus, but all tests (including this last one of yours) indicate around a 30% improvement.

facug91 commented:
Thanks @pfeatherstone for all the testing and @davisking for the suggestions, I think it's ready to review :)

facug91 commented:
Sorry, do not accept this PR yet please, I need to run more tests. Tomorrow when I finish I will publish the results.

facug91 commented:
Well, I did a lot of testing and I think the C++ optimizer played a trick on us here. In tests where the same frame is processed over and over again, they are optimized in a weird way by the compiler and are not accurate.

This is the script we were using above, with frames being resized to 1920x1080 resolution:

```cpp
#include <opencv2/opencv.hpp>
#include <dlib/matrix.h>
#include <dlib/opencv.h>
#include <dlib/image_transforms/assign_image.h>

using namespace dlib;
using namespace std::chrono;

void assign_ptr(matrix<unsigned char>& dst, const cv::Mat& src) {
	if (dst.nr() != src.rows || dst.nc() == src.cols)
		dst.set_size(src.rows, src.cols);
	const long size = dst.nr() * dst.nc();
	std::memcpy(&dst(0,0), src.data, size);
}

int main(int argc, char** argv)
{
	cv::VideoCapture cap("/home/investigacion/dep/dlib/dlib/test/ffmpeg_data/MOT17-13-SDP-raw.h265");
	cv::Mat frame, gray;
	cap >> frame;
	cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
	cv::resize(gray, gray, cv::Size(1080, 1920));

	matrix<unsigned char> aux(gray.rows, gray.cols);

	const int ntests = 500;

	{
		const auto start = high_resolution_clock::now();
		for (int i = 0 ; i < ntests ; ++i)
			assign_image(aux, dlib::cv_image<unsigned char>(gray));
		std::cout << "assign_image : " << duration_cast<microseconds>(high_resolution_clock::now() - start).count() / ntests << " us" << std::endl;
		const long ans = sum(aux);
		std::cout << "Sum of mats  : " << ans << std::endl;
	}

	{
		const auto start = high_resolution_clock::now();
		for (int i = 0 ; i < ntests ; ++i)
			aux = mat(gray.data, gray.rows, gray.cols);
		std::cout << "mat          : " << duration_cast<microseconds>(high_resolution_clock::now() - start).count() / ntests << " us" << std::endl;
		const long ans = sum(aux);
		std::cout << "Sum of mats  : " << ans << std::endl;
	}

	{
		const auto start = high_resolution_clock::now();
		for (int i = 0 ; i < ntests ; ++i)
			assign_ptr(aux, gray);
		std::cout << "assign_ptr   : " << duration_cast<microseconds>(high_resolution_clock::now() - start).count() / ntests << " us" << std::endl;
		const long ans = sum(aux);
		std::cout << "Sum of mats  : " << ans << std::endl;
	}

	return 0;
}
```

The following tests were executed in full hd (1920x1080), with different values for ntests (number of iterations):

```
With 500 iterations
assign_image :   99 us
mat          :   95 us
assign_ptr   :   68 us

With 50 iterations
assign_image :  270 us
mat          :  118 us
assign_ptr   :   85 us

With 34 iterations
assign_image :  120 us
mat          :   97 us
assign_ptr   :   69 us

With 33 iterations
assign_image : 2292 us
mat          : 1138 us
assign_ptr   :   69 us

With 10 iterations
assign_image : 2396 us
mat          : 1159 us
assign_ptr   :   85 us

With 5 iterations
assign_image : 2404 us
mat          : 1139 us
assign_ptr   :   82 us

With 2 iterations
assign_image : 2627 us
mat          : 1151 us
assign_ptr   :  104 us
```

And on top of that, the test results also depend on the order. If I swap `assign_image` and `assign_ptr`, these are some of the results:

```
With 34 iterations
assign_ptr   :  110 us
mat          :  100 us
assign_image :   95 us

With 10 iterations
assign_ptr   :  313 us
mat          : 1170 us
assign_image : 2244 us

With 2 iterations
assign_ptr   :  630 us
mat          : 1278 us
assign_image : 2388 us
```

So I decided to test them individually (as I did at the beginning), but with a single script, with parameters to choose the method and resolution:

```cpp
#include <opencv2/opencv.hpp>
#include <dlib/matrix.h>
#include <dlib/opencv.h>
#include <dlib/image_transforms/assign_image.h>

using namespace dlib;
using namespace std::chrono;

void assign_ptr(matrix<unsigned char>& dst, const cv_image<unsigned char>& src) {
	if (dst.nr() != src.nr() || dst.nc() == src.nc())
		dst.set_size(src.nr(), src.nc());
	const long size = dst.nr() * dst.nc();
	std::memcpy(&dst(0, 0), &src(0, 0), size);
}

int main(int argc, char** argv) {

	if (argc != 4) {
		std::cout << "This script needs 3 parameters: method code to test (1: assign_image, 2: mat, 3: assign_ptr), "
		             "rows and columns. e.g. 1 720 1280, to test assign_image in HD resolution";
		return -1;
	}

	int code = std::atoi(argv[1]);
	int rows = std::atoi(argv[2]);
	int cols = std::atoi(argv[3]);

	cv::VideoCapture cap("/home/investigacion/dep/dlib/dlib/test/ffmpeg_data/MOT17-13-SDP-raw.h265");
	cv::Mat frame, gray;
	cap >> frame;

	long long counter = 0, ans = 0, frames = 0;
	matrix<unsigned char> aux(rows, cols);

	cap >> frame;
	while (!frame.empty()) {
		cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
		cv::resize(gray, gray, cv::Size(rows, cols));

		if (code == 1){
			const auto start = high_resolution_clock::now();
			assign_image(aux, dlib::cv_image<unsigned char>(gray));
			const auto end = high_resolution_clock::now();
			counter += duration_cast<microseconds>(end - start).count();
		}
		else
		if (code == 2){
			const auto start = high_resolution_clock::now();
			aux = mat(gray.data, gray.rows, gray.cols);
			const auto end = high_resolution_clock::now();
			counter += duration_cast<microseconds>(end - start).count();
		}
		else
		if (code == 3) {
			const auto start = high_resolution_clock::now();
			assign_ptr(aux, gray);
			const auto end = high_resolution_clock::now();
			counter += duration_cast<microseconds>(end - start).count();
		}
		else {
			std::cout << "Wrong code provided." << std::endl;
			return -1;
		}

		ans += dlib::sum(aux);
		frames++;
		cap >> frame;
	}
	std::cout << "Resolution " << gray.size() << std::endl;
	if (code == 1) std::cout << "assign_image avg time: " << counter / frames << " us, with sum " << ans << std::endl;
	if (code == 2) std::cout << "mat          avg time: " << counter / frames << " us, with sum " << ans << std::endl;
	if (code == 3) std::cout << "assign_ptr   avg time: " << counter / frames << " us, with sum " << ans << std::endl;

	return 0;
}
```

Then I execute all tests at the same time with the following:

```bash
./runner 1 360 640 && \
./runner 2 360 640 && \
./runner 3 360 640 && \
./runner 1 720 1280 && \
./runner 2 720 1280 && \
./runner 3 720 1280 && \
./runner 1 1080 1920 && \
./runner 2 1080 1920 && \
./runner 3 1080 1920 && \
./runner 1 2160 3840 && \
./runner 2 2160 3840 && \
./runner 3 2160 3840 && \
./runner 1 4320 7680 && \
./runner 2 4320 7680 && \
./runner 3 4320 7680
```

Results in my computer (i5-9400f)

```
Resolution [360 x 640]
assign_image avg time:   250 us, with sum 94994
mat          avg time:   128 us, with sum 94994
assign_ptr   avg time:     8 us, with sum 94994
Resolution [720 x 1280]
assign_image avg time:   977 us, with sum 94720
mat          avg time:   508 us, with sum 94720
assign_ptr   avg time:    42 us, with sum 94720
Resolution [1080 x 1920]
assign_image avg time:  2181 us, with sum 96652
mat          avg time:  1135 us, with sum 96652
assign_ptr   avg time:   102 us, with sum 96652
Resolution [2160 x 3840]
assign_image avg time:  8768 us, with sum 96099
mat          avg time:  4612 us, with sum 96099
assign_ptr   avg time:   553 us, with sum 96099
Resolution [4320 x 7680]
assign_image avg time: 35106 us, with sum 96046
mat          avg time: 18398 us, with sum 96046
assign_ptr   avg time:  2289 us, with sum 96046
```

Results in a Jetson Nano

```
Resolution [360 x 640]
assign_image avg time:   2145 us, with sum 98102
mat          avg time:   1946 us, with sum 98102
assign_ptr   avg time:     88 us, with sum 98102
Resolution [720 x 1280]
assign_image avg time:   8249 us, with sum 96526
mat          avg time:   7581 us, with sum 96526
assign_ptr   avg time:    343 us, with sum 96526
Resolution [1080 x 1920]
assign_image avg time:  18472 us, with sum 95038
mat          avg time:  16983 us, with sum 95038
assign_ptr   avg time:    701 us, with sum 95038
Resolution [2160 x 3840]
assign_image avg time:  73421 us, with sum 93851
mat          avg time:  67911 us, with sum 93851
assign_ptr   avg time:   2508 us, with sum 93851
Resolution [4320 x 7680]
assign_image avg time: 293340 us, with sum 93914
mat          avg time: 270648 us, with sum 93914
assign_ptr   avg time:   9177 us, with sum 93914
```

If you see something that I did wrong, or can run the same tests to see the results on other computers, that would be great. But if this is accurate enough, the performance gains are bigger than we thought a couple of days ago, especially on low-power CPUs like the Jetson Nanos.

davisking commented:
Yeah you have the right of it. 

The code is a bit off, as you said though.  I put the right way to handle all the cases in https://github.com/davisking/dlib/pull/2826.  Try that out and let me know if it's faster for you and then I'll merge it in :)

facug91 commented:
Yeah, a bit off, and I didn't take into account for those cases with stride and non trivially copyable types, my apologies.
I hope you don't mind, but I copied your code here and added an optimization for matrices with stride. Both cases run faster on my machine. I've even tried an edge case of a matrix with 2000000 rows, 1 column, and a stride of 2, and still much faster.

davisking commented:
Cool, yeah it's fine.  I just pushed a change and once the tests run I'll merge it.  Thanks for pointing out the speed problem :)
